### PR TITLE
feat(security): add rate limiting

### LINKED ISSUE #15: Add rate limiting to authentication endpoints
## Overview
Add rate limiting to protect authentication endpoints from brute-force and abuse attacks.

## Audit Findings (2026-02-13)
Security audit confirmed: **zero rate limiting exists** on any endpoint. 5 rapid registration requests all completed successfully at ~250ms each with no throttling or `X-RateLimit-*` headers. This was rated **HIGH severity**.

### Verified Attack Vectors
- Brute-force password guessing (no lockout after failed attempts)
- Registration spam / mass account creation (confirmed: all attempts succeed)
- Credential stuffing attacks
- Denial-of-wallet (Firebase charges per auth operation)

### 1. Install slowapi (FastAPI rate limiting library)
```bash
uv add slowapi
```

### 2. Configure rate limiter in `app/__init__.py`

Redis is **required** as the rate limiting storage backend (in-memory storage is not acceptable in production — counters would reset on every restart and not be shared across workers).

```python
from slowapi import Limiter, _rate_limit_exceeded_handler
from slowapi.util import get_remote_address
from slowapi.errors import RateLimitExceeded
from app.config.redis import get_redis_config

redis_config = get_redis_config()
limiter = Limiter(key_func=get_remote_address, storage_uri=redis_config["url"])
app.state.limiter = limiter
app.add_exception_handler(RateLimitExceeded, _rate_limit_exceeded_handler)
```

### 3. Apply rate limits to endpoints
```python
from app import limiter
from fastapi import Request

@router.post("/register")
@limiter.limit("5/hour")
async def register(request: Request, body: RegisterRequest):
    ...
```

### Rate Limits
| Endpoint | Limit | Rationale |
|----------|-------|-----------|
| `POST /v1/auth/register` | 5/hour per IP | Prevent registration spam |
| `POST /v1/auth/login` | 10/15minute per IP | Prevent brute-force password guessing |
| `GET /v1/movies/search` | 30/minute per IP | Prevent API abuse |
| `POST /v1/logs/` | 20/minute per IP | Prevent spam |

### Files to Modify
- `pyproject.toml` - Add slowapi dependency (via `uv add slowapi`)
- `app/__init__.py` - Initialize rate limiter with Redis storage URI
- `app/controllers/auth_controller.py` - Apply rate limits to register and login
- `app/controllers/log_controller.py` - Apply rate limits
- `app/controllers/movie_controller.py` - Apply rate limits

## Acceptance Criteria
- [ ] Rate limiting library installed and configured
- [ ] Registration endpoint limited to 5 requests/hour per IP
- [ ] Login endpoint limited to 10 requests/15 minutes per IP
- [ ] Rate limit exceeded returns 429 status code with `Retry-After` header
- [ ] `X-RateLimit-Remaining` and `X-RateLimit-Limit` headers included in responses
- [ ] Unit tests for rate limiting behavior
- [ ] Redis backend used for rate limiting (required — `REDIS_ENABLED` flag removed, Redis is always on)

### PR DESCRIPTION
**Pull request by benincasantonio**

### CODE REVIEW

**Review by benincasantonio — COMMENTED** [review]
Review from Claude — a few things worth checking before merge. Overall the layered IP + session + per-account design looks solid; comments inline.

**Comment by benincasantonio** [comment]
Follow-up from the review discussion: the silent IP fallback in `get_rate_limit_key` (when `request.state.user_id` isn't populated) is the kind of bug that's invisible until a shared-IP customer complains. Two lightweight mitigations worth adding before or shortly after this merges:

1. **Log the fallback.** In `app/utils/rate_limit_utils.py`, when `user_id` is `None` on a route that went through `auth_dependency`, emit a warning (or bump a `rate_limit_key_type{type="ip"}` counter). Any auth'd route leaking into the IP branch then shows up immediately instead of silently throttling unrelated users sharing a NAT.

2. **Invariant test.** One parametrized test that iterates `app.routes`, calls each auth'd endpoint with a valid session, and asserts the resulting slowapi Redis key is prefixed with `user:` (not `ip:`). Catches the whole class of regressions — including anyone who later copy-pastes a custom `key_func` onto an authenticated route — without needing per-endpoint assertions.

Same logging idea would also catch the empty `JWT_SECRET_KEY` fallback flagged in `auth_rate_limit_service.py`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/services/auth_rate_limit_service.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    LOGIN_ACCOUNT_RATE_LIMIT_SCOPE,
+    LOGIN_FAILED_ACCOUNT_LIMIT_ITEM,
+    RESET_PASSWORD_ACCOUNT_LIMIT_ITEM,
+    RESET_PASSWORD_ACCOUNT_RATE_LIMIT_SCOPE,
+)
+from app.config.rate_limiter import limiter
+from app.utils.error_codes import ErrorCodes
+from app.utils.exceptions import AppException
+
+_RATE_LIMIT_HMAC_SECRET = os.getenv("JWT_SECRET_KEY", "")
```
Comment (benincasantonio) [inline app/services/auth_rate_limit_service.py]: **Silent empty-secret fallback.** `os.getenv("JWT_SECRET_KEY", "")` is evaluated at import time and defaults to an empty string. If the env var is missing (misconfigured deploy, test bootstrap order), every account key is HMAC'd with a zero-length key — still deterministic, so rate limiting keeps working, but the failure is completely silent and the keyspace is now predictable by anyone who knows the algorithm.

Two concerns:
1. Consider raising at startup if the secret is empty, or reading it lazily via a config accessor that validates.
2. Reusing `JWT_SECRET_KEY` for a second purpose couples rate-limit bucket identity to JWT signing. Rotating the JWT key would orphan every in-flight account lockout bucket (attackers get a free reset). A dedicated `RATE_LIMIT_HMAC_SECRET` would be cleaner.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/controllers/auth_controller.py [hunk 2]
```diff
@@ -28,25 +39,45 @@
...
+        user = await auth_service.login(
+            email=request_body.email, password=request_body.password
+        )
+    except AppException as exc:
+        if exc.error == ErrorCodes.INVALID_CREDENTIALS:
+            auth_rate_limit_service.register_login_failure(request_body.email)
+        raise
 
     # Set Cookies
     set_auth_cookies(response, str(user.id))
```
Comment (benincasantonio) [inline app/controllers/auth_controller.py:87]: **Successful login doesn't clear the per-account failure counter.** `register_login_failure` is only called on `INVALID_CREDENTIALS`, but there's no corresponding clear on success. A legitimate user who mistypes their password 4 times, logs in successfully, then mistypes once more within the 15-minute window gets locked out on the 5th cumulative failure even though they already proved ownership.

Suggest calling `auth_rate_limit_service` to clear the `LOGIN_ACCOUNT_RATE_LIMIT_SCOPE` bucket for this email right after `set_auth_cookies`. The service already has `_clear_limit` — just expose a `clear_login_failures(email)` method.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: app/middleware/rate_limit_session_middleware.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                setattr(request.state, RATE_LIMIT_SESSION_STATE, session_id)
+                await self._rate_limit_cache.upsert_session(session_id)
+                has_valid_session = True
+
+        if not has_valid_session:
+            issued_session_id = generate_rate_limit_session_id()
+            setattr(request.state, RATE_LIMIT_SESSION_STATE, issued_session_id)
+            await self._rate_limit_cache.upsert_session(issued_session_id)
+            if session_id is not None:
+                request.cookies[RATE_LIMIT_SESSION_COOKIE] = issued_session_id
```
Comment (benincasantonio) [inline app/middleware/rate_limit_session_middleware.py]: **This mutation is dead code.** `request.cookies[RATE_LIMIT_SESSION_COOKIE] = issued_session_id` mutates the parsed cookie dict on the request, but nothing downstream re-reads it — `get_session_rate_limit_key` reads from `request.state.{RATE_LIMIT_SESSION_STATE}`, which was already set on line 42. Either drop the assignment or add a comment explaining which consumer depends on it. As-is it's misleading.

---
_Generated by [Claude Code](https://claude.ai/code)_
